Sources/NIOCore/NIOSplitLinesMessageDecoder.swift

```diff
@@ -0,0 +1,414 @@
+                return slice
+            }
+
+            // "\r\n" is 2 bytes long, so we need to manually account for it.
+            switch separator {
+            case UInt8(ascii: "\n") where slice.readableBytes == 0:
+                let isCRLF = self.previousSeparatorWasCR
+                self.previousSeparatorWasCR = false
+                if isCRLF {
+                    continue
+                }
+            case UInt8(ascii: "\r"):
+                self.previousSeparatorWasCR = true
+            default:
+                self.previousSeparatorWasCR = false
+            }
+
+            return slice
+        }
+    }
+
+    /// Decode the next message separated by one of the ASCII line breaks.
+    /// To be used when we're still receiving data.
+    @inlinable
+    public mutating func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+        try self.decode(buffer: &buffer, hasReceivedLastChunk: false)
+    }
+
+    /// Decode the next message separated by one of the ASCII line breaks.
+    /// To be used when the last chunk of data has been received.
+    @inlinable
+    public mutating func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> InboundOut? {
+        try self.decode(buffer: &buffer, hasReceivedLastChunk: true)
+    }
+}
+
+@available(*, unavailable)
+extension NIOSplitLinesMessageDecoder: Sendable {}
+
+// MARK: - NIOSplitUTF8LinesMessageDecoder
+
+/// A decoder which splits the data into subsequences that are separated by line breaks.
+///
+/// Use `AsyncSequence/splitLines(omittingEmptySubsequences:maximumBufferSize:)`
+/// or `AsyncSequence/splitUTF8Lines(omittingEmptySubsequences:maximumBufferSize:)` to create a
+/// `NIODecodedAsyncSequence` that uses this decoder.
+///
+/// The following Characters are considered line breaks, similar to
+/// standard library's `String.split(whereSeparator: \.isNewline)`:
+/// - "\n" (U+000A): LINE FEED (LF)
+/// - U+000B: LINE TABULATION (VT)
+/// - U+000C: FORM FEED (FF)
+/// - "\r" (U+000D): CARRIAGE RETURN (CR)
+/// - "\r\n" (U+000D U+000A): CR-LF
+///
+/// The following Characters are NOT considered line breaks, unlike in
+/// standard library's `String.split(whereSeparator: \.isNewline)`:
+/// - U+0085: NEXT LINE (NEL)
+/// - U+2028: LINE SEPARATOR
+/// - U+2029: PARAGRAPH SEPARATOR
+///
+/// This is because these characters would require unicode and data-encoding awareness, which
+/// are outside swift-nio's scope.
+///
+/// Usage:
+/// ```swift
+/// let baseSequence = MyAsyncSequence<ByteBuffer>(...)
+/// let splitLinesSequence = baseSequence.splitUTF8Lines()
+///
+/// for try await string in splitLinesSequence {
+///     print("Split by line breaks!\n", string)
+/// }
+/// ```
+public struct NIOSplitUTF8LinesMessageDecoder: NIOSingleStepByteToMessageDecoder {
+    public typealias InboundOut = String
+
+    @usableFromInline
+    var splitLinesDecoder: NIOSplitLinesMessageDecoder
+
+    @inlinable
+    init(omittingEmptySubsequences: Bool) {
```

Comment:
Should this be `public` given the type is?

Reply:
Done